[Feature]: node bindings for linux-(ppc64,s390x)-gnu

### What problem does this feature solve?

We (Red Hat) run builds with rspack in a number of architectures including `x64`, `arm64`, `s390x`, and `ppc64`. Unfortunately there is no native API binding support for ppc64 or s390x yet, so running builds on these architectures produce errors: 

```
time="2026-07-27T06:13:26Z" level=info msg="buildah [stderr] /opt/app-root/src/frontend/node_modules/@rspack/binding/binding.js:395"
time="2026-07-27T06:13:26Z" level=info msg="buildah [stderr]     throw new Error("
time="2026-07-27T06:13:26Z" level=info msg="buildah [stderr]           ^"
time="2026-07-27T06:13:26Z" level=info msg="buildah [stderr] Error: Cannot find native binding. npm has a bug related to optional dependencies (https://github.com/npm/cli/issues/4828). Please try `npm i` again after removing both package-lock.json and node_modules directory."
time="2026-07-27T06:13:26Z" level=info msg="buildah [stderr] "
time="2026-07-27T06:13:26Z" level=info msg="buildah [stderr] Cannot find module './rspack.linux-s390x-gnu.node'"
time="2026-07-27T06:13:26Z" level=info msg="buildah [stderr] Require stack:"
time="2026-07-27T06:13:26Z" level=info msg="buildah [stderr] - /opt/app-root/src/frontend/node_modules/@rspack/binding/binding.js"
time="2026-07-27T06:13:26Z" level=info msg="buildah [stderr] Cannot find module '@rspack/binding-linux-s390x-gnu'"
```

```
time="2026-07-27T06:10:36Z" level=info msg="buildah [stderr] /opt/app-root/src/frontend/node_modules/@rspack/binding/binding.js:395"
time="2026-07-27T06:10:36Z" level=info msg="buildah [stderr]     throw new Error("
time="2026-07-27T06:10:36Z" level=info msg="buildah [stderr]           ^"
time="2026-07-27T06:10:36Z" level=info msg="buildah [stderr] Error: Cannot find native binding. npm has a bug related to optional dependencies (https://github.com/npm/cli/issues/4828). Please try `npm i` again after removing both package-lock.json and node_modules directory."
time="2026-07-27T06:10:36Z" level=info msg="buildah [stderr] "
time="2026-07-27T06:10:36Z" level=info msg="buildah [stderr] Cannot find module './rspack.linux-ppc64-gnu.node'"
time="2026-07-27T06:10:36Z" level=info msg="buildah [stderr] Require stack:"
time="2026-07-27T06:10:36Z" level=info msg="buildah [stderr] - /opt/app-root/src/frontend/node_modules/@rspack/binding/binding.js"
time="2026-07-27T06:10:36Z" level=info msg="buildah [stderr] Cannot find module '@rspack/binding-linux-ppc64-gnu'"
```

### What does the proposed API of configuration look like?

New `optionalDependencies` for `@rspack/binding` on npm: `@rspack/binding-linux-ppc64-gnu`, `@rspack/binding-linux-s390x-gnu`, and maybe `@rspack/binding-linux-s390x-musl`/`@rspack/binding-linux-s390x-musl` if those make sense too.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -145,6 +145,26 @@ jobs:
       profile: 'ci'
       runner: ${{ vars.LINUX_SELF_HOSTED_RUNNER_LABELS || '"ubuntu-22.04"' }}
 
+  build-ppc64:
+    name: Build ppc64
+    needs: [check-changed]
+    if: ${{ needs.check-changed.outputs.code_changed == 'true' }}
+    uses: ./.github/workflows/reusable-build-build.yml
+    with:
+      target: powerpc64le-unknown-linux-gnu
+      profile: 'ci'
+      runner: ${{ vars.LINUX_SELF_HOSTED_RUNNER_LABELS || '"ubuntu-22.04"' }}
+
+  build-s390x:
+    name: Build s390x
+    needs: [check-changed]
+    if: ${{ needs.check-changed.outputs.code_changed == 'true' }}
+    uses: ./.github/workflows/reusable-build-build.yml
+    with:
+      target: s390x-unknown-linux-gnu
+      profile: 'ci'
+      runner: ${{ vars.LINUX_SELF_HOSTED_RUNNER_LABELS || '"ubuntu-22.04"' }}
+
   # TODO: Enable this job after Rspack portable cache is implemented.
   check-cache:
     name: Check Cache Portable
@@ -187,6 +207,8 @@ jobs:
         test-mac,
         build-wasm,
         build-riscv64,
+        build-ppc64,
+        build-s390x,
         test-wasm,
         check-changed,
         size-limit,
@@ -200,13 +222,13 @@ jobs:
       - name: Test check
         if: ${{ needs.check-changed.outputs.code_changed == 'true'
           && github.event_name == 'pull_request'
-          && join(needs.*.result, ',')!='success,success,success,success,success,success,success,success,success,skipped' }}
+          && join(needs.*.result, ',')!='success,success,success,success,success,success,success,success,success,success,success,skipped' }}
         run: echo "Test Failed" && exit 1
 
       - name: Test check
         if: ${{ needs.check-changed.outputs.code_changed == 'true'
           && github.event_name != 'pull_request'
-          && join(needs.*.result, ',')!='success,success,success,success,success,success,success,success,skipped,skipped' }}
+          && join(needs.*.result, ',')!='success,success,success,success,success,success,success,success,success,success,skipped,skipped' }}
         run: echo "Test Failed" && exit 1
 
       - name: No check to Run test

--- a/.github/workflows/preview-commit.yml
+++ b/.github/workflows/preview-commit.yml
@@ -50,6 +50,10 @@ jobs:
             runner: ${{ needs.get-runner-labels.outputs.LINUX_RUNNER_LABELS }}
           - target: riscv64gc-unknown-linux-musl
             runner: "'ubuntu-latest'"
+          - target: powerpc64le-unknown-linux-gnu
+            runner: "'ubuntu-latest'"
+          - target: s390x-unknown-linux-gnu
+            runner: "'ubuntu-latest'"
           - target: i686-pc-windows-msvc
             runner: ${{ needs.get-runner-labels.outputs.WINDOWS_RUNNER_LABELS }}
           - target: x86_64-pc-windows-msvc

--- a/.github/workflows/release-canary.yml
+++ b/.github/workflows/release-canary.yml
@@ -52,6 +52,10 @@ jobs:
             runner: ${{ vars.LINUX_SELF_HOSTED_RUNNER_LABELS || '"ubuntu-22.04"' }}
           - target: riscv64gc-unknown-linux-musl
             runner: ${{ vars.LINUX_SELF_HOSTED_RUNNER_LABELS || '"ubuntu-22.04"' }}
+          - target: powerpc64le-unknown-linux-gnu
+            runner: ${{ vars.LINUX_SELF_HOSTED_RUNNER_LABELS || '"ubuntu-22.04"' }}
+          - target: s390x-unknown-linux-gnu
+            runner: ${{ vars.LINUX_SELF_HOSTED_RUNNER_LABELS || '"ubuntu-22.04"' }}
           - target: i686-pc-windows-msvc
             runner: ${{ vars.WINDOWS_SELF_HOSTED_RUNNER_LABELS || '"windows-latest"' }}
           - target: x86_64-pc-windows-msvc

--- a/.github/workflows/reusable-build-all.yml
+++ b/.github/workflows/reusable-build-all.yml
@@ -37,6 +37,10 @@ jobs:
             runner: ${{ vars.LINUX_SELF_HOSTED_RUNNER_LABELS || '"ubuntu-22.04"' }}
           - target: riscv64gc-unknown-linux-musl
             runner: ${{ vars.LINUX_SELF_HOSTED_RUNNER_LABELS || '"ubuntu-22.04"' }}
+          - target: powerpc64le-unknown-linux-gnu
+            runner: ${{ vars.LINUX_SELF_HOSTED_RUNNER_LABELS || '"ubuntu-22.04"' }}
+          - target: s390x-unknown-linux-gnu
+            runner: ${{ vars.LINUX_SELF_HOSTED_RUNNER_LABELS || '"ubuntu-22.04"' }}
           - target: i686-pc-windows-msvc
             runner: ${{ '"windows-latest"' }}
           - target: x86_64-pc-windows-msvc

--- a/.github/workflows/reusable-build-build.yml
+++ b/.github/workflows/reusable-build-build.yml
@@ -86,6 +86,10 @@ jobs:
         with:
           tool: cargo-zigbuild
 
+      - name: Install clang with PowerPC and SystemZ backends
+        if: ${{ inputs.target == 'powerpc64le-unknown-linux-gnu' || inputs.target == 's390x-unknown-linux-gnu' }}
+        run: command -v clang-15 || { sudo apt-get update && sudo apt-get install -y clang-15; }
+
       # setup rust target for native runner
       - name: Setup Rust Target
         run: rustup target add ${{ inputs.target }}
@@ -155,13 +159,13 @@ jobs:
         if: ${{ inputs.target == 'powerpc64le-unknown-linux-gnu' }}
         run: RUST_TARGET=powerpc64le-unknown-linux-gnu USE_NAPI_CROSS=1 pnpm build:binding:${{ inputs.profile }}
         env:
-          TARGET_CC: clang
+          TARGET_CC: clang-15
 
       - name: Build s390x-unknown-linux-gnu
         if: ${{ inputs.target == 's390x-unknown-linux-gnu' }}
         run: RUST_TARGET=s390x-unknown-linux-gnu USE_NAPI_CROSS=1 pnpm build:binding:${{ inputs.profile }}
         env:
-          TARGET_CC: clang
+          TARGET_CC: clang-15
 
       # Windows
       - name: Build i686-pc-windows-msvc

--- a/.github/workflows/reusable-build-build.yml
+++ b/.github/workflows/reusable-build-build.yml
@@ -151,6 +151,18 @@ jobs:
         env:
           TARGET_CC: clang
 
+      - name: Build powerpc64le-unknown-linux-gnu
+        if: ${{ inputs.target == 'powerpc64le-unknown-linux-gnu' }}
+        run: RUST_TARGET=powerpc64le-unknown-linux-gnu USE_NAPI_CROSS=1 pnpm build:binding:${{ inputs.profile }}
+        env:
+          TARGET_CC: clang
+
+      - name: Build s390x-unknown-linux-gnu
+        if: ${{ inputs.target == 's390x-unknown-linux-gnu' }}
+        run: RUST_TARGET=s390x-unknown-linux-gnu USE_NAPI_CROSS=1 pnpm build:binding:${{ inputs.profile }}
+        env:
+          TARGET_CC: clang
+
       # Windows
       - name: Build i686-pc-windows-msvc
         if: ${{ inputs.target == 'i686-pc-windows-msvc' }}

--- a/.github/workflows/reusable-build-test.yml
+++ b/.github/workflows/reusable-build-test.yml
@@ -97,6 +97,8 @@ jobs:
         node: ${{ fromJSON(
           inputs.target == 'wasm32-wasip1-threads'&& '[24]'
           || inputs.target == 'riscv64gc-unknown-linux-gnu' && '[22]'
+          || inputs.target == 'powerpc64le-unknown-linux-gnu' && '[22]'
+          || inputs.target == 's390x-unknown-linux-gnu' && '[22]'
           || contains(inputs.target, 'linux') && '[22, 24]'
           || '[22]') }}
     name: Test Node ${{ matrix.node }}

--- a/.github/workflows/reusable-release-npm.yml
+++ b/.github/workflows/reusable-release-npm.yml
@@ -58,6 +58,12 @@ jobs:
           - target: riscv64gc-unknown-linux-musl
             runner: ${{ vars.LINUX_SELF_HOSTED_RUNNER_LABELS || '"ubuntu-22.04"' }}
             test-runner: '"ubuntu-22.04"'
+          - target: powerpc64le-unknown-linux-gnu
+            runner: ${{ vars.LINUX_SELF_HOSTED_RUNNER_LABELS || '"ubuntu-22.04"' }}
+            test-runner: '"ubuntu-22.04"'
+          - target: s390x-unknown-linux-gnu
+            runner: ${{ vars.LINUX_SELF_HOSTED_RUNNER_LABELS || '"ubuntu-22.04"' }}
+            test-runner: '"ubuntu-22.04"'
           - target: i686-pc-windows-msvc
             runner: ${{ vars.WINDOWS_SELF_HOSTED_RUNNER_LABELS || '"windows-latest"' }}
             test-runner: '"windows-latest"'

--- a/.gitignore
+++ b/.gitignore
@@ -155,6 +155,8 @@ npm/**/*.node
 !npm/linux-arm64-musl
 !npm/linux-riscv64-gnu
 !npm/linux-riscv64-musl
+!npm/linux-ppc64-gnu
+!npm/linux-s390x-gnu
 !npm/linux-x64-gnu/
 !npm/linux-x64-musl
 !npm/wasm32-wasi

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -58,7 +58,7 @@ insta               = { version = "1.42.0", default-features = false }
 itertools           = { version = "0.14.0", default-features = false, features = ["use_std"] }
 itoa                = { version = "1.0.18", default-features = false }
 json                = { version = "0.12.4", default-features = false }
-json-escape-simd    = { version = "3.1.0", default-features = false }
+json-escape-simd    = { version = "3.1.1", default-features = false }
 lightningcss        = { version = "1.0.0-alpha.71", default-features = false, features = ["serde"] }
 md4                 = { version = "0.10.2", default-features = false }
 memchr              = { version = "2.8.2", default-features = false }

--- a/crates/node_binding/package.json
+++ b/crates/node_binding/package.json
@@ -57,7 +57,9 @@
       "aarch64-unknown-linux-musl",
       "aarch64-pc-windows-msvc",
       "riscv64gc-unknown-linux-gnu",
-      "riscv64gc-unknown-linux-musl"
+      "riscv64gc-unknown-linux-musl",
+      "powerpc64le-unknown-linux-gnu",
+      "s390x-unknown-linux-gnu"
     ],
     "wasm": {
       "initialMemory": 16384,
@@ -76,6 +78,8 @@
     "@rspack/binding-linux-arm64-musl": "workspace:*",
     "@rspack/binding-linux-riscv64-gnu": "workspace:*",
     "@rspack/binding-linux-riscv64-musl": "workspace:*",
+    "@rspack/binding-linux-ppc64-gnu": "workspace:*",
+    "@rspack/binding-linux-s390x-gnu": "workspace:*",
     "@rspack/binding-linux-x64-gnu": "workspace:*",
     "@rspack/binding-linux-x64-musl": "workspace:*",
     "@rspack/binding-wasm32-wasi": "workspace:*",

--- a/crates/rspack_binding_builder_testing/package.json
+++ b/crates/rspack_binding_builder_testing/package.json
@@ -45,7 +45,9 @@
       "aarch64-unknown-linux-musl",
       "aarch64-pc-windows-msvc",
       "riscv64gc-unknown-linux-gnu",
-      "riscv64gc-unknown-linux-musl"
+      "riscv64gc-unknown-linux-musl",
+      "powerpc64le-unknown-linux-gnu",
+      "s390x-unknown-linux-gnu"
     ],
     "wasm": {
       "initialMemory": 16384,

--- a/npm/linux-ppc64-gnu/package.json
+++ b/npm/linux-ppc64-gnu/package.json
@@ -1,0 +1,29 @@
+{
+  "name": "@rspack/binding-linux-ppc64-gnu",
+  "version": "2.1.5",
+  "license": "MIT",
+  "description": "Node binding for rspack",
+  "main": "rspack.linux-ppc64-gnu.node",
+  "homepage": "https://rspack.rs",
+  "bugs": "https://github.com/web-infra-dev/rspack/issues",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/web-infra-dev/rspack",
+    "directory": "packages/rspack"
+  },
+  "publishConfig": {
+    "access": "public",
+    "os": [
+      "linux"
+    ],
+    "cpu": [
+      "ppc64"
+    ],
+    "libc": [
+      "glibc"
+    ]
+  },
+  "files": [
+    "rspack.linux-ppc64-gnu.node"
+  ]
+}

--- a/npm/linux-s390x-gnu/package.json
+++ b/npm/linux-s390x-gnu/package.json
@@ -1,0 +1,29 @@
+{
+  "name": "@rspack/binding-linux-s390x-gnu",
+  "version": "2.1.5",
+  "license": "MIT",
+  "description": "Node binding for rspack",
+  "main": "rspack.linux-s390x-gnu.node",
+  "homepage": "https://rspack.rs",
+  "bugs": "https://github.com/web-infra-dev/rspack/issues",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/web-infra-dev/rspack",
+    "directory": "packages/rspack"
+  },
+  "publishConfig": {
+    "access": "public",
+    "os": [
+      "linux"
+    ],
+    "cpu": [
+      "s390x"
+    ],
+    "libc": [
+      "glibc"
+    ]
+  },
+  "files": [
+    "rspack.linux-s390x-gnu.node"
+  ]
+}

--- a/scripts/build-npm.cjs
+++ b/scripts/build-npm.cjs
@@ -12,6 +12,8 @@ const CpuToNodeArch = {
   i686: 'ia32',
   armv7: 'arm',
   riscv64gc: 'riscv64',
+  powerpc64le: 'ppc64',
+  s390x: 's390x',
 };
 
 const SysToNodePlatform = {

--- a/website/docs/en/contribute/development/releasing.md
+++ b/website/docs/en/contribute/development/releasing.md
@@ -19,6 +19,8 @@ During the release, the following binary artifacts for the target platforms are 
 - x86_64-unknown-linux-gnu
 - aarch64-unknown-linux-gnu
 - riscv64gc-unknown-linux-gnu
+- powerpc64le-unknown-linux-gnu
+- s390x-unknown-linux-gnu
 - x86_64-unknown-linux-musl
 - aarch64-unknown-linux-musl
 - riscv64gc-unknown-linux-musl

--- a/website/docs/en/guide/start/quick-start.mdx
+++ b/website/docs/en/guide/start/quick-start.mdx
@@ -46,6 +46,8 @@ Currently supported platforms:
 - `linux-arm64-musl`
 - `linux-riscv64-gnu`
 - `linux-riscv64-musl`
+- `linux-ppc64-gnu`
+- `linux-s390x-gnu`
 - `linux-x64-gnu`
 - `linux-x64-musl`
 - `win32-arm64-msvc`

--- a/website/docs/zh/contribute/development/releasing.md
+++ b/website/docs/zh/contribute/development/releasing.md
@@ -19,6 +19,8 @@ Latest 是最新的稳定版本，遵循 Semantic Versioning 语义化版本号�
 - x86_64-unknown-linux-gnu
 - aarch64-unknown-linux-gnu
 - riscv64gc-unknown-linux-gnu
+- powerpc64le-unknown-linux-gnu
+- s390x-unknown-linux-gnu
 - x86_64-unknown-linux-musl
 - aarch64-unknown-linux-musl
 - riscv64gc-unknown-linux-musl

--- a/website/docs/zh/guide/start/quick-start.mdx
+++ b/website/docs/zh/guide/start/quick-start.mdx
@@ -46,6 +46,8 @@ Rspack 支持使用 [Node.js](https://nodejs.org/)、[Deno](https://deno.com/) �
 - `linux-arm64-musl`
 - `linux-riscv64-gnu`
 - `linux-riscv64-musl`
+- `linux-ppc64-gnu`
+- `linux-s390x-gnu`
 - `linux-x64-gnu`
 - `linux-x64-musl`
 - `win32-arm64-msvc`

--- a/website/project-words.txt
+++ b/website/project-words.txt
@@ -152,6 +152,8 @@ perfetto
 Peschke
 pftrace
 pmmmwh
+powerpc64le
+ppc64
 proxied
 proxying
 px2rem
@@ -180,6 +182,7 @@ ruleoneof
 ruleparserdataurlcondition
 Rust
 rustup
+s390x
 samply
 samsung
 sanyuan

